README.md
# tap-exchangeratesapi

A [Singer](https://singer.io) Tap to extract currency exchange rate
data from [exchangeratesapi.io](http://exchangeratesapi.io).

---

Copyright &copy; 2017 Stitch

### config.json:
```
{
    "base": "AUD",
    "start_date": "2017-02-02",
    "api_key": "123abc456def"
}

```

### state.json
```
{"start_date": "2021-03-31"}
```

## Installation
### Prepare Server

1. Installation of Ubuntu 18.04
2. Install python 3.7.1 from source code
```
sudo apt update
sudo apt install build-essential zlib1g-dev libncurses5-dev libgdbm-dev libnss3-dev libssl-dev libreadline-dev libffi-dev wget
cd /tmp
wget https://www.python.org/ftp/python/3.7.1/Python-3.7.1.tgz
tar –xf Python-3.7.1.tgz
```

Info: https://phoenixnap.com/kb/how-to-install-python-3-ubuntu

3. Install pip3
See here: https://linuxize.com/post/how-to-install-pip-on-ubuntu-18.04/

4. Install setuptools for pip
```
pip install setuptools
```

### Create python environments
We use different environments for each pymodul from singer.io to avoid version conflicts
1. Create tap-exchangeratesapi environment
```
#Create environment
python3 -m venv .virtualenvs/tap-exchangeratesapi
#Activate environment
source .virtualenvs/tap-exchangeratesapi/bin/activate
# Go to exchangeratesapi repo
cd tap-exchangeratesapi
#Install requierements
pip install -r requirements.txt
# Install tap-exchangeratesapi as py modul in environment
python setup.py build --force
python setup.py install
```

2. Create environment for target for example stitch or csv
```
# Install target-stitch in its own virtualenv
python3 -m venv .virtualenvs/target-stitch
source .virtualenvs/target-stitch/bin/activate
pip install target-stitch
deactivate


# Install target-csv in its own virtualenv
python3 -m venv .virtualenvs/target-csv
source .virtualenvs/target-csv/bin/activate
pip install target-csv
deactivate
```

## EXECUTE THE PACKAGE (with a target)

.virtualenvs/tap-exchangeratesapi/bin/tap-exchangeratesapi --config "config_fil" --state "state-file" | .virtualenvs/target-csv/bin/target-csv

CHANGELOG.md
# Changelog
## 0.1.2
  * Dependabot update [#15](https://github.com/singer-io/tap-exchangeratesapi/pull/15)


tap_exchangeratesapi.py
#!/usr/bin/env python3

import json
import sys
import argparse
import time
import requests
import singer
import backoff
import copy

from datetime import date, datetime, timedelta

base_url = 'https://api.exchangeratesapi.io/'

logger = singer.get_logger()
session = requests.Session()

DATE_FORMAT='%Y-%m-%d'

def parse_response(r):
    flattened = r['rates']
    flattened[r['base']] = 1.0
    flattened['date'] = time.strftime('%Y-%m-%dT%H:%M:%SZ', time.strptime(r['date'], DATE_FORMAT))
    return flattened

schema = {'type': 'object',
          'properties':
          {'date': {'type': 'string',
                    'format': 'date-time'}}}

def giveup(error):
    logger.error(error.response.text)
    response = error.response
    return not (response.status_code == 429 or
                response.status_code >= 500)

@backoff.on_exception(backoff.constant,
                      (requests.exceptions.RequestException),
                      jitter=backoff.random_jitter,
                      max_tries=5,
                      giveup=giveup,
                      interval=30)
def request(url, params):
    response = requests.get(url=url, params=params)
    response.raise_for_status()
    return response
    
def do_sync(base, start_date):
    logger.info('Replicating exchange rate data from exchangeratesapi.io starting from {}'.format(start_date))

    state = {'start_date': start_date}
    next_date = start_date
    prev_schema = {}
    
    try:
        while datetime.strptime(next_date, DATE_FORMAT) <= datetime.utcnow():
            response = request(base_url + next_date, {'base': base})
            payload = response.json()

            # Update schema if new currency/currencies exist
            for rate in payload['rates']:
                if rate not in schema['properties']:
                    schema['properties'][rate] = {'type': ['null', 'number']}

            # Only write schema if it has changed
            if schema != prev_schema:
                singer.write_schema('exchange_rate', schema, 'date')

            if payload['date'] == next_date:
                singer.write_records('exchange_rate', [parse_response(payload)])

            state = {'start_date': next_date}
            next_date = (datetime.strptime(next_date, DATE_FORMAT) + timedelta(days=1)).strftime(DATE_FORMAT)
            prev_schema = copy.deepcopy(schema)

    except requests.exceptions.RequestException as e:
        logger.fatal('Error on ' + e.request.url +
                     '; received status ' + str(e.response.status_code) +
                     ': ' + e.response.text)
        singer.write_state(state)
        sys.exit(-1)

    singer.write_state(state)
    logger.info('Tap exiting normally')


def main():
    parser = argparse.ArgumentParser()

    parser.add_argument(
        '-c', '--config', help='Config file', required=False)
    parser.add_argument(
        '-s', '--state', help='State file', required=False)

    args = parser.parse_args()

    if args.config:
        with open(args.config) as file:
            config = json.load(file)
    else:
        config = {}

    if args.state:
        with open(args.state) as file:
            state = json.load(file)
    else:
        state = {}

    start_date = state.get('start_date') or config.get('start_date') or datetime.utcnow().strftime(DATE_FORMAT)
    start_date = singer.utils.strptime_with_tz(start_date).date().strftime(DATE_FORMAT)

    do_sync(config.get('base', 'USD'), start_date)


if __name__ == '__main__':
    main()


tap_exchangeratesapi/discovery.py
"""Discovery functions for the Singer.io tap"""

import json
import os

import singer
from singer import Catalog, metadata

from typing import Dict, List, Any

LOGGER = singer.get_logger()
REPLICATION_METHOD = "INCREMENTAL"


def _get_abs_path(path):
    return os.path.join(os.path.dirname(os.path.realpath(__file__)), path)


# Load schemas from schemas folder
def _load_stream_schemas():
    schemas = {}
    for filename in os.listdir(_get_abs_path("schemas")):
        path = _get_abs_path("schemas") + "/" + filename
        basename = filename.replace(".json", "")
        with open(path) as file:
            schemas[basename] = json.load(file)
    return schemas


def discover() -> Dict[str, Dict[str, Any]]:
    """
    Run discovery

    Returns
    -------
    Dict[str, Dict[str, Any]]
        The list of streams with their associated metadata
    """
    LOGGER.info("Starting discovery mode")
    streams = []
    for stream_name, stream in _load_stream_schemas().items():
        catalog_entry = {
            "stream": stream_name,
            "tap_stream_id": stream_name,
            "schema": stream["schema"],
            "metadata": metadata.get_standard_metadata(
                schema=stream["schema"],
                key_properties=["date"],
                valid_replication_keys=["date"],
                replication_method=REPLICATION_METHOD,
            ),
            "key_properties": ["date"],
        }
        streams.append(catalog_entry)
    return Catalog.from_dict({"streams": streams})


tap_exchangeratesapi/__init__.py
#!/usr/bin/env python3

import json
import sys
import argparse
import time
import requests
import singer
import backoff
import copy

from datetime import date, datetime, timedelta

base_url = 'https://api.apilayer.com/exchangerates_data/'

logger = singer.get_logger()
session = requests.Session()

DATE_FORMAT='%Y-%m-%d'

SCHEMA = {'type': 'object',
          'properties':
            {'date': {'type': 'string',
                    'format': 'date-time'},
             'rates': {'type': 'string'}
             }
          }

def parse_response(r):
    #flattened = r['rates']
    flattened = {}
    flattened['base'] = r['base']
    flattened['rates'] = json.dumps(r['rates'])
    flattened['date'] = time.strftime('%Y-%m-%dT%H:%M:%SZ', time.strptime(r['date'], DATE_FORMAT))
    return flattened
def giveup(error):
    logger.error(error.response.text)
    response = error.response
    return not (response.status_code == 429 or
                response.status_code >= 500)
@backoff.on_exception(backoff.constant,
                      (requests.exceptions.RequestException),
                      jitter=backoff.random_jitter,
                      max_tries=5,
                      giveup=giveup,
                      interval=30)
def request(url, params, headers):
    response = requests.get(url=url, params=params,headers=headers)
    response.raise_for_status()
    return response
    
def do_sync(base, start_date, apikey):
    state = {'start_date': start_date}
    next_date = start_date
    # prev_schema = {}
    singer.write_schema('exchange_rate', SCHEMA, 'date')
    
    try:
        while datetime.strptime(next_date, DATE_FORMAT) <= datetime.utcnow():
            logger.info('Replicating exchange rate data from %s using base %s with apikey %s',
                        next_date,
                        base,
                        apikey)

            response = request(base_url + next_date, {'base': base}, {'apikey':apikey})
            payload = response.json()

            # # Update schema if new currency/currencies exist
            # for rate in payload['rates']:
            #     if rate not in schema['properties']:
            #         schema['properties'][rate] = {'type': ['null', 'number']}
            #
            # # Only write schema if it has changed
            # if schema != prev_schema:
            #     singer.write_schema('exchange_rate', schema, 'date')

            if payload['date'] == next_date:
                singer.write_records('exchange_rate', [parse_response(payload)])

            state = {'start_date': next_date}
            next_date = (datetime.strptime(next_date, DATE_FORMAT) + timedelta(days=1)).strftime(DATE_FORMAT)
            # prev_schema = copy.deepcopy(schema)

    except requests.exceptions.RequestException as e:
        logger.fatal('Error on ' + e.request.url +
                     '; received status ' + str(e.response.status_code) +
                     ': ' + e.response.text)
        singer.write_state(state)
        sys.exit(-1)

    singer.write_state(state)
    logger.info('Tap exiting normally')


def main():
    parser = argparse.ArgumentParser()

    parser.add_argument(
        '-c', '--config', help='Config file', required=False)
    parser.add_argument(
        '-s', '--state', help='State file', required=False)

    args = parser.parse_args()

    if args.config:
        with open(args.config) as file:
            config = json.load(file)
    else:
        config = {}

    if args.state:
        with open(args.state) as file:
            state = json.load(file)
    else:
        state = {}

    start_date = state.get('start_date') or config.get('start_date') or datetime.utcnow().strftime(DATE_FORMAT)
    start_date = singer.utils.strptime_with_tz(start_date).date().strftime(DATE_FORMAT)

    do_sync(config.get('base', 'USD'), start_date, config.get('api_key',''))


if __name__ == '__main__':
    main()
